p4-16/spec/P4-16-spec.mdk

```diff
@@ -1180,7 +1180,113 @@ produced by different programmers at different times:
+  P4 compiler for `import`-ed modules.  In general, the compiler will
+  maintain an ordered list of module library paths, and every `import`
+  statement will search the specified files in order under each of the
+  library paths, loading the first file that matches.
+
+There are also some important differences between modules and
+`#included` files:
+
+* `#include` directives produce flat files.  In contrast, P4 modules
+  can optionally create new namespaces, allowing P4 programs to make
+  use of multiple libraries without concern about name clashes.
+
+* `#include` directives do not understand anything about the P4 syntax
+  and semantics, they are purely string processing operations.  In
+  particular, the included files do not have to be syntactically
+  correct P4 programs or program fragments.
+
+* `#include` directives function transitively: a file included from
+  another included file has the exact same visibility as the top-level
+  constructs.  In contrast, transitively included modules are
+  completely invisible at the top-level.
+
+The following keywords are used to construct modules in P4:
+
+* `import`, `into`: import definitions from a module
+* `private`: mark definitions of a module as non-importable
+
+### Importing modules { #sec-import }
+
+A definitions in a module can be imported into the global namespace
+with the following statement, executed at compile-time.
+
+`import module/path/file.p4`
+
+~ Begin P4Grammar
+importStatement: IMPORT filePath
+               | IMPORT filePath INTO namespacePath
+               ;
+
+filePath: FILENAME
+        | DIRECTORYNAME SLASH filePath
+        ;
+
+namespacePath: IDENTIFIER
+             | namespacePath DOT IDENTIFIER
+             ;
+~ End P4Grammar
+
+`import` statements must be present at the top-level only.  The effect
+of the `import` statement is to include all the definitions in the
+imported module in the place of the `import` module.  An import is
+followed by an optional path separated by slashes and by a file name.
+Only *relative* paths are allowed.
+
+Alternatively, the definitions in a module can be imported in a
+separate namespace with the following construct:
+
+`import module/path/file.p4 into namespace`
+
+The effect of this statement is similar to the simple import; in
+addition, the names of all declarations that are imported will be
+prefixed with the `namespace.` prefix (including a trailing dot).
+
+If a module is imported into a namespace then all its declarations are
+available prefixed with the respective namespace.
+
+~ Begin P4Grammar
+pathPrefix: DOT
+          | IDENTIFIER DOT
+          | pathPrefix IDENTIFIER DOT
+          ;
+~ End P4Grammar
+
+### Private declarations
+
+A declaration in a module can be prefixed with the `private` keyword.
+This will cause the declaration not to be visible in the importing
+programs; however, the declaration is still accessible in the imported
+code.
```

Comment:
Maybe everything should be `private` by default -- add a `public` or `export` keyword and get rid of `private`

Reply:
I thought about that, the downside is that people may need to modify libraries they have written so far using `#include` to add `export` to each declaration. So I thought that `public` by default is simpler. But for a clean design I would prefer private by default.